src/apis/Event.re

```diff
@@ -1,106 +1,143 @@
-  "target": Js.Nullable.t(float),
-  "timeStamp": float,
-  "_type": Js.Nullable.t(string),
+module EventMethods = (T: {
+                         type target('a);
+                         type r;
+                       }) => {
+  [@bs.get] [@bs.return nullable]
+  external bubbles: T.target('a) => option(bool) = "bubbles";
+  [@bs.get] [@bs.return nullable]
+  external cancelable: T.target('a) => option(bool) = "cancelable";
+  [@bs.get] external currentTarget: T.target('a) => float = "currentTarget";
+  [@bs.get] [@bs.return nullable]
+  external defaultPrevented: T.target('a) => option(bool) =
+    "defaultPrevented";
+  [@bs.get] external dispatchConfig: T.target('a) => T.r = "dispatchConfig";
+  [@bs.get] [@bs.return nullable]
+  external eventPhase: T.target('a) => option(float) = "eventPhase";
+  [@bs.send] external preventDefault: T.target('a) => unit = "preventDefault";
+  [@bs.send]
+  external isDefaultPrevented: T.target('a) => bool = "isDefaultPrevented";
+  [@bs.send]
+  external stopPropagation: T.target('a) => unit = "stopPropagation";
+  [@bs.send]
+  external isPropagationStopped: T.target('a) => bool =
+    "isPropagationStopped";
+  [@bs.get] [@bs.return nullable]
+  external isTrusted: T.target('a) => option(bool) = "isTrusted";
+  [@bs.get] external nativeEvent: T.target('a) => 'a = "nativeEvent";
+  [@bs.send] external persist: T.target('a) => unit = "persist";
+  [@bs.get] [@bs.return nullable]
+  external target: T.target('a) => option(float) = "target";
+  [@bs.get] external timeStamp: T.target('a) => float = "timeStamp";
+  [@bs.get] [@bs.return nullable]
+  external _type: T.target('a) => option(string) = "type";
 };
 
-type responderSyntheticEvent('a) = {
-  .
-  // synthethicEvent keys
-  "bubbles": Js.Nullable.t(bool),
-  "cancelable": Js.Nullable.t(bool),
-  "currentTarget": float,
-  "defaultPrevented": Js.Nullable.t(bool),
-  "dispatchConfig": {. "registrationName": string},
-  "eventPhase": Js.Nullable.t(float),
-  [@bs.meth] "preventDefault": unit => unit,
-  [@bs.meth] "isDefaultPrevented": unit => bool,
-  [@bs.meth] "stopPropagation": unit => unit,
-  [@bs.meth] "isPropagationStopped": unit => bool,
-  "isTrusted": Js.Nullable.t(bool),
-  "nativeEvent": 'a,
-  [@bs.meth] "persist": unit => unit,
-  "target": Js.Nullable.t(float),
-  "timeStamp": float,
-  "_type": Js.Nullable.t(string),
-  // responderSyntheticEvent additional key
-  "touchHistory": {
-    .
-    "indexOfSingleActiveTouch": float,
-    "mostRecentTimeStamp": float,
-    "numberActiveTouches": float,
-    "touchBank":
-      array({
-        .
-        "touchActive": bool,
-        "startPageX": float,
-        "startPageY": float,
-        "startTimeStamp": float,
-        "currentPageX": float,
-        "currentPageY": float,
-        "currentTimeStamp": float,
-        "previousPageX": float,
-        "previousPageY": float,
-        "previousTimeStamp": float,
-      }),
-  },
+module SyntheticEvent = {
+  type t_('a) = {
```

Comment:
Why `t_` and not `t`? Can the typename be more descriptive if it cannot be `t`?

Reply:
When `SyntheticEvent` is included in some event module, `type t` would also be included and the compiler won't allow types to be redefined. I couldn't come up with a better name, so suggestions are welcome.

Since `SyntheticEvent` and `ResponderSyntheticEvent` need to be used to define new events, we can add two modules (such as `_SyntheticEvent` and `_ResponderSyntheticEvent`)  for all modules to inherit from. Then in `SyntheticEvent` and `ResponderSyntheticEvent` we could have `type t ('a) = t_('a)`.

Reply:
That would allow users to define new events either by inhering from `_SyntheticEvent` or using `SyntheticEvent.t('payload)` without a new module.